newIDE/app/src/MainFrame/EditorTabs/DraggableEditorTabs.js

```diff
@@ -0,0 +1,155 @@
+// @flow
+import * as React from 'react';
+
+import { makeDragSourceAndDropTarget } from '../../UI/DragAndDrop/DragSourceAndDropTarget';
+import { ScreenTypeMeasurer } from '../../UI/Reponsive/ScreenTypeMeasurer';
+import { ColumnDropIndicator } from './DropIndicator';
+import {
+  type EditorTabsState,
+  type EditorTab,
+  getEditors,
+  getCurrentTabIndex,
+} from './EditorTabsHandler';
+import {
+  ClosableTabs,
+  ClosableTab,
+  type ClosableTabProps,
+} from '../../UI/ClosableTabs';
+
+type DraggableClosableTabsProps = {|
+  hideLabels?: boolean,
+  editorTabs: EditorTabsState,
+  onClickTab: (index: number) => void,
+  onCloseTab: (editor: EditorTab) => void,
+  onCloseOtherTabs: (editor: EditorTab) => void,
+  onCloseAll: () => void,
+  onTabActive: (editor: EditorTab) => void,
+  onDropTab: (fromIndex: number, toIndex: number) => void,
+|};
+
+export function DraggableClosableTabs({
+  hideLabels,
+  editorTabs,
+  onClickTab,
+  onCloseTab,
+  onCloseOtherTabs,
+  onCloseAll,
+  onTabActive,
+  onDropTab,
+}: DraggableClosableTabsProps) {
+  let draggedTabIndex: number | typeof undefined = undefined;
```

Comment:
Prefer using the hook useState:
```js
const [draggedTabIndex, setDraggedTabIndex] = React.useState<?number>(null);
```

Reply:
Using state triggers a rerender when `onBeginDrag()` is called making us lose the drag, so it is not possible

Reply:
If this is just a local variable, then it's fine. Though:

```suggestion
  let draggedTabIndex: number | undefined = undefined;
```

or better:

```suggestion
  let draggedTabIndex: ?number = null;
```

In most cases, avoid "undefined". 

Undefined could means "there is not even a value here", like a parameter "was not even passed". It's possibly a sign of an error.
While null means "This value is defined, but it's defined to be null, surely because it's represents a state where it has no valid value". With `null`, we can clearly see that you meant what you wrote, why seeing "undefined" while debugging is much more worrying and hard to interpret because it might means that you passed something that is not even existing.

null make it clear that it's, well, a null value (and not a value that is not even defined)

Reply:
As discussed with @fannieyan, using a state variable here triggers a re-render of the component when a tab is picked and dragged and it messes with react-dnd. Maybe there's something wrong in the way the drag n drop was implemented because I can't see a way in which you can have a state variable but no re-render when it changes @4ian  @ClementPasteau